x/gamm/keeper/invariants.go

```diff
@@ -3,149 +3,49 @@ package keeper
-// 	}
-// }
+func AllInvariants(keeper Keeper, bk types.BankKeeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		msg, broke := PoolAccountInvariant(keeper, bk)(ctx)
+		return msg, broke
+	}
+}
 
 // PoolAccountInvariant checks that the pool account balance reflects the sum of
 // pool assets
-// func PoolAccountInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant {
-// 	return func(ctx sdk.Context) (string, bool) {
-// 		pools, err := keeper.GetPools(ctx)
-// 		if err != nil {
-// 			return sdk.FormatInvariant(types.ModuleName, poolBalanceInvariantName,
-// 				"\tgamm pool retrieval failed"), true
-// 		}
-
-// 		for _, pool := range pools {
-// 			assetCoins := types.PoolAssetsCoins(pool.GetAllPoolAssets())
-// 			accCoins := bk.GetAllBalances(ctx, pool.GetAddress())
-// 			if !assetCoins.IsEqual(accCoins) {
-// 				return sdk.FormatInvariant(types.ModuleName, poolBalanceInvariantName,
-// 					fmt.Sprintf("\tgamm pool id %d\n\tasset coins: %s\n\taccount coins: %s\n",
-// 						pool.GetId(), assetCoins, accCoins)), true
-// 			}
-// 		}
-
-// 		return sdk.FormatInvariant(types.ModuleName, poolBalanceInvariantName,
-// 			"\tgamm all pool asset coins and account coins match\n"), false
-// 	}
-// }
-
-// PoolTotalWeightInvariant checks that the pool total weight reflect the sum of
-// pool weights
-// func PoolTotalWeightInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant {
-// 	return func(ctx sdk.Context) (string, bool) {
-// 		pools, err := keeper.GetPools(ctx)
-// 		if err != nil {
-// 			return sdk.FormatInvariant(types.ModuleName, "pool-total-weight",
-// 				"\tgamm pool retrieval failed"), true
-// 		}
-
-// 		for _, pool := range pools {
-// 			totalWeight := sdk.ZeroInt()
-// 			for _, asset := range pool.GetAllPoolAssets() {
-// 				totalWeight = totalWeight.Add(asset.Weight)
-// 			}
-// 			if !totalWeight.Equal(pool.GetTotalWeight()) {
-// 				return sdk.FormatInvariant(types.ModuleName, "pool-total-weight",
-// 					fmt.Sprintf("\tgamm pool id %d\n\tcalculated weight sum %s\n\tpool total weight: %s\n",
-// 						pool.GetId(), totalWeight, pool.GetTotalWeight())), true
-// 			}
-// 		}
-
-// 		return sdk.FormatInvariant(types.ModuleName, "pool-total-weight",
-// 			"\tgamm all pool calculated and stored total weight match\n"), false
-// 	}
-// }
-
-func genericPow(base, exp sdk.Dec) sdk.Dec {
-	if !base.GTE(sdk.NewDec(2)) {
-		return osmomath.Pow(base, exp)
+func PoolAccountInvariant(keeper Keeper, bk types.BankKeeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		pools, err := keeper.GetPools(ctx)
+		if err != nil {
+			return sdk.FormatInvariant(types.ModuleName, poolBalanceInvariantName,
+				"\tgamm pool retrieval failed"), true
+		}
+
+		for _, pool := range pools {
+			expectedCoins := pool.GetTotalPoolLiquidity(ctx)
+			actualCoins := bk.GetAllBalances(ctx, pool.GetAddress())
+			if !actualCoins.IsAllGTE(expectedCoins) {
```

Comment:
These three lines changed from the commented out code that existed before. Basically it checks that the balance thats thought to be in the module account, is actually there when you look at the bank module.

Reply:
Before it wanted exact equality, which had issues. (other processes could give coins to the module account)